flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/FsSavepointStore.java

```diff
@@ -0,0 +1,225 @@
+
+	private static final Logger LOG = LoggerFactory.getLogger(FsSavepointStore.class);
+
+	/** Magic number for sanity checks against stored savepoints. */
+	int MAGIC_NUMBER = 0x4960672d;
+
+	/** Root path for savepoints. */
+	private final Path rootPath;
+
+	/** Prefix for savepoint files. */
+	private final String prefix;
+
+	/** File system to use for file access. */
+	private final FileSystem fileSystem;
+
+	/**
+	 * Creates a new file system based {@link SavepointStore}.
+	 *
+	 * @param rootPath Root path for savepoints
+	 * @param prefix   Prefix for savepoint files
+	 * @throws IOException On failure to access root path
+	 */
+	FsSavepointStore(String rootPath, String prefix) throws IOException {
+		this.rootPath = new Path(checkNotNull(rootPath, "Root path"));
+		this.prefix = checkNotNull(prefix, "Prefix");
+
+		this.fileSystem = FileSystem.get(this.rootPath.toUri());
+	}
+
+	@Override
+	public <T extends Savepoint> String storeSavepoint(T savepoint) throws IOException {
+		Preconditions.checkNotNull(savepoint, "Savepoint");
+
+		Exception latestException = null;
+		Path path = null;
+		FSDataOutputStream fdos = null;
+
+		// Try to create a FS output stream
+		for (int attempt = 0; attempt < 10; attempt++) {
+			path = new Path(rootPath, FileUtils.getRandomFilename(prefix));
+			try {
+				fdos = fileSystem.create(path, false);
+				break;
+			} catch (Exception e) {
+				latestException = e;
+			}
+		}
+
+		if (fdos == null) {
+			throw new IOException("Failed to create file output stream at " + path, latestException);
+		}
+
+		boolean success = false;
+		try (DataOutputStream dos = new DataOutputStream(fdos)) {
+			// Write header
+			dos.writeInt(MAGIC_NUMBER);
+			dos.writeInt(savepoint.getVersion());
+
+			// Write savepoint
+			SavepointSerializer<T> serializer = SavepointSerializers.getSerializer(savepoint);
+			serializer.serialize(savepoint, fdos);
+			success = true;
+		} finally {
+			fdos.close();
+
+			if (!success && fileSystem.exists(path)) {
+				if (!fileSystem.delete(path, true)) {
+					LOG.warn("Failed to delete file " + path + " after failed write.");
+				}
+			}
+		}
+
+		return path.toString();
+	}
+
+	@Override
+	public Savepoint loadSavepoint(String path) throws IOException {
+		Preconditions.checkNotNull(path, "Path");
+
+		try {
+			try (FSDataInputStream fdis = createFsInputStream(new Path(path))) {
+				try (DataInputStream dis = new DataInputStream(fdis)) {
+					int magicNumber = dis.readInt();
+					int version = dis.readInt();
+
+					if (magicNumber == MAGIC_NUMBER) {
+						SavepointSerializer<?> serializer = SavepointSerializers.getSerializer(version);
+						Savepoint savepoint = serializer.deserialize(fdis);
+						return savepoint;
+					} else {
+						throw new IllegalStateException("Unexpected magic number. This indicates " +
+								"that the specified file is not a proper savepoint or " +
+								"the file has been corrupted.");
+					}
+				}
+			}
+		} catch (Throwable t) {
```

Comment:
Exceptions which are thrown due to a wrong path or because the input stream couldn't be opened will also trigger the savepoint version 0 loading. Maybe we could differentiate between unable to open file stream and wrong magic number or unable to read version.